### ISSUE TITLE
Bold formatting bugs - need jquery.autotype to write a test

### ISSUE THREAD

**Issue by jywarren** [publiclab/PublicLab.Editor#307]
**Please note we are preparing to participate in Google Code-in, and have reserved this issue for GCI participants - but we'd love to have your help with another one! Please check out https://code.publiclab.org to see more.**

Problem statement by the user Jeffrey: I was typing into an editor, and did the following sequence:

1. made a newline (new paragraph)
2. pressed `ctrl-b` (or used the bold button)
3. typed `Hello`, replacing the `strong text` example text that appeared when i pressed bold
4. pressed `ctrl-b` again to /unbold/, saw `strong text` again but not bolded this time
5. typed `, world` replacing the 2nd (not bold) instance of `strong text`
6. went back to the start of the line, before `Hello` and pressed enter to add a line break
7. started typing (it was bold)
8. pressed `ctrl-b` to unbold and got `strong text` but bold this time
9. typed ` world` and pressed `ctrl-b` again, got unbolded `strong text`.
10. pressed Markdown mode to see what was going on

I did this a couple times, this was the output in the most recent:

```

  
****

****Hello**, there**strong text******Hello **world, 
```

I think really we need to tackle this in smaller parts. I suggest that we install the `autotype` plugin so that we can simulate typing here: https://github.com/mmonteleone/jquery.autotype

Then we can try to reproduce a subset of the above sequence in a test,

**Comment by SidharthBansal** [comment]
PUBLISHED to GCI dashboard

**Comment by jywarren** [comment]
Hmm. Was this fixed recently, i thought perhaps?

**Comment by jywarren** [comment]
Maybe linked to this? https://github.com/publiclab/PublicLab.Editor/pull/220

**Comment by keshav234156** [comment]
@jywarren I was just thinking of rough idea of how we can solve this . The problem is occurring because of https://github.com/bevacqua/woofmark/blob/master/src/html/wrapping.js#L19-L21 .Only the HTML mode is using the wrapping function. The Markdown mode is using other function and it works perfectly in Markdown mode(Please check it so that I am confirmed that it's the correct behavior). We can directly make changes to https://github.com/bevacqua/woofmark , but author doesn't seem to be active(as the previous commit was 1 year ago). What we can do is that,to  write our own bold and wrapping function similar to https://github.com/publiclab/PublicLab.Editor/blob/master/src/modules/PublicLab.RichTextModule.HorizontalRule.js .Src code would be similar to https://github.com/bevacqua/woofmark but with the desired modification which solves this bug.
What do you think about this Or do you have any other workaround in your mind?

**Comment by keshav234156** [comment]
@jywarren We could then also be able to implement the feature of the darkening of the icon when bold is on like we have in google docs.

The basic idea is to introduce woofmark in our src code rather than having a node module So tat we can format it according to our needs.

**Comment by keshav234156** [comment]
Italic text also faces similar issue

**Comment by jywarren** [comment]
Looks great, good analysis and great noting it affects italics too! Thanks!

**Comment by shreyaa-s** [comment]
I agree with @keshav234156 , we should write our own wrapping function. I've been tracking down a bug #459 and the implementation of `wrapping` seems to have a lot of discrepancies. I think we can start with discussing what the wrapping function is supposed to do, making sure that we don't miss out any feature, and then we can look at what it actually does right now. The comparison would make it easier and clear as to what we need to add or implement.

**Comment by shreyaa-s** [comment]
From what I've worked with so far, the current implementation includes:
1. Adding the necessary tags, (opening) before and (closing)after the selected text. Pointing it out that these additions are not made to the selected text but rather to the text before and after the selected text.
2. Closing any tag that is already opened in selected text and opening a tag that is closed in selected text. These changes are made to the selected text.
 For better understanding have a look at  https://github.com/publiclab/PublicLab.Editor/issues/457#issuecomment-627813960
